Trim output for maps

bartekgorny opened the pull request:
Similar principle as in recon_rec applied to maps.

ferd commented on src/recon_map.erl:
```diff
@@ -0,0 +1,171 @@
+
+map_matches(#{} = M, Pattern) when is_function(Pattern) ->
+    Pattern(M);
+map_matches(#{} = M, #{} = Pattern) ->
```
I like the `fun(M)` pattern since it ought to be decently fast, but once again, very nervous about the form below.

I guess at least if the map has to have both the keys and the values it makes sense, but I'm kind of wondering why you'd set this filter and not instead do it in the pattern of the match specification for the trace in the first place? Is this only intended for output filtering? Because input filtering would be far more effective by applying it to the trace pattern directly.

ferd commented on src/recon_map.erl:
```diff
@@ -0,0 +1,171 @@
+apply_map_limits(all, _) ->
+    #{};
+apply_map_limits(Fields, M) ->
+    maps:with(Fields, M).
```
Alright, here I see that the main point is really to constrain anything that matched previously to only output _some_ fields. So this wouldn't be usable to generically strip out PII, only to help dig better into something you know you want to look into.

ferd commented on src/recon_rec.erl:
```diff
@@ -54,11 +55,12 @@ is_active() ->
 %% @doc remove all imported definitions, destroy the table, clean up
 clear() ->
     maybe_kill(recon_ets),
+    recon_map:clear(),
```
why are you clearing maps in the rec module?

ferd commented on src/recon_rec.erl:
```diff
@@ -80,25 +82,29 @@ list() ->
-    end.
+    end;
+limit(_, _, _) ->
+    {error, "Bad argument - the spec is limit(atom(), integer(), limit())"}.
```
let it crash, document the fields instead.

ferd commented on src/recon_rec.erl:
```diff
@@ -80,25 +82,29 @@ list() ->
 %% @private if a tuple is a known record, formats is as "#recname{field=value}", otherwise returns
 %% just a printout of a tuple.
+format_tuple(Tuple) when element(1, Tuple) == set ->
+    ["set:", recon_trace:format_trace_output(true, sets:to_list(Tuple))];
```
this is majorly risky and I disagree with this approach. This will cause failures when trying to format _any_ tuple that has the first element `set` and is not actually the `set` data structure. This has to go.

ferd commented on src/recon_trace.erl:
```diff
@@ -607,36 +607,49 @@ to_hms(_) ->
+    [Label1,
+     "#{",
+        join(", ", [format_kv(Recs, Key, Val) || {Key, Val} <- ItemList]),
+    "}"];
```
wait, so this label is made to look like a record? So if I have a filter with label `user_map` then I get the output `user_map#{...}`? That's confusing.

ferd replied:
Oh and it uses format_kvs which makes it impossible to distinguish between a map and a record in trace output, too.

I think this is a loss of resolution that risks impeding debugging more than helping it.

ferd commented on src/recon_map.erl:
```diff
@@ -0,0 +1,171 @@
+map_matches(#{} = M, Pattern) when is_function(Pattern) ->
+    Pattern(M);
+map_matches(#{} = M, #{} = Pattern) ->
+    map_matches(M, maps:to_list(Pattern));
```
This should be done when saving the pattern, no way this is worth doing on every single map during an active trace

bartekgorny commented:
Ok, sorry about that - I'll work on addressing your concerns. About the background - I'm a developer of MongooseIM, and internally it uses an 'accumulator' which is a big map. Any tracing immediately clutters the screen. So the rationale behind this is the same as in case of formatting records - to be able to selectively limit the output. The trouble with maps, though, is that while a record can be almost uniquely identified by name and arity, a map is, well, just a map. I then had to come up with some way of identifying them, hence the 'patterns'. For example, our 'acc' always has mongoose_acc=>true, so if I define a one-key map #{mongoose_acc=>true} as the pattern, it will be recognised, apply limits to the matched map and leave everything else intact. This is what these patterns are for, plus you can write your fun if you need more elaborate matching. I actually use it daily and it proves to be very useful, but I admit there are weak points, thanks for pointing them out.

ferd commented:
So the objective is more about removing huge maps from the traces than only displaying some of them. When you debug, are you looking for specific fields in there, or mostly just looking to remove the huge map from the output? That's the kind of stuff about the debugging approach that would help narrow down the solution more strictly.

bartekgorny commented:
The objective is to display only selected fields from these huge maps. The "patterns" are sort of a typing system, bolted on to recognise which maps are these "huge" ones I want to preprocess, and add a type tag to the output. I'm not a big fan of maps, for exactly this reason - they are all one type.

bartekgorny commented:
I think I addressed all comments.

About performance - I did some load testing, see https://gist.github.com/bartekgorny/04a14e579812c037177d8b8a9ffb2d60.
They showed no impact on performance. Let me know if you believe some heavier/more elaborate load testing is needed. I could think of some sophisticated pattern indexing system (to avoid iterating through all pattern list every time), though based on what I found out so far there doesn't seem to be a dire need of it.

ferd commented on src/recon_map.erl:
```diff
@@ -0,0 +1,191 @@
+-spec limit(map_label(), pattern(), limit()) -> ok | {error, any()}.
+limit(Label, #{} = Pattern, Limit) when is_atom(Label) ->
+    store_pattern(Label, Pattern, Limit);
+limit(Label, Pattern, Limit) when is_atom(Label), is_function(Pattern) ->
```
This clause can never work because there's a maps:to_list conversion done in the storing function

ferd commented:
Sorry for the delay in reviewing. I'll be merging this once the final little issue in the limits function is fixed, the rest is good to go.

ferd commented on src/recon_map.erl:
```diff
@@ -0,0 +1,198 @@
+    ok.
+
+prepare_limit(Limit) when is_binary(Limit) -> [Limit];
+prepare_limit(Limit) when is_atom(Limit) -> [Limit];
```
This messes up with `all` and `none` now.
